python/cudf/cudf/core/udf/groupby_typing.py

```diff
@@ -153,15 +176,86 @@ def caller(data, index, size):
     call_cuda_functions[funcname.lower()][type_key] = caller
 
 
-def _make_unary_attr(funcname):
-    class GroupUnaryReductionAttrTyping(AbstractTemplate):
-        key = f"GroupType.{funcname}"
+class GroupOpBase(AbstractTemplate):
+    def make_error_string(self, args):
+        fname = self.key.__name__
+        sr_err = ", ".join(["Series" for _ in range(len(args))])
+        return (
+            f"{fname}({sr_err}) is not supported by JIT GroupBy "
+            f"apply. Supported features are listed at:\n{_UDF_DOC_URL}"
+        )
+
+    def generic(self, args, kws):
+        # early exit to make sure typing doesn't fail for normal
+        # non-group ops
+        if not all(isinstance(arg, GroupType) for arg in args):
+            return None
+        # check if any groups are poisoned for this op
+        for arg in args:
+            if isinstance(arg.group_scalar_type, types.Poison):
+                raise UDFError(
+                    f"Use of a column of {arg.group_scalar_type.ty} detected "
+                    "within UDF body. \nOnly columns of the following dtypes "
+                    "may be used through the \nGroupBy.apply() JIT engine: "
+                    f"{[str(x) for x in SUPPORTED_GROUPBY_NUMPY_TYPES]}"
+                )
+        fname = self.key.__name__
+        if fname in call_cuda_functions:
+            for sig in call_cuda_functions[fname].keys():
+                if all(
+                    arg.group_scalar_type == ty for arg, ty in zip(args, sig)
+                ):
+                    return nb_signature(sig[0], *args)
```

Comment:
Good place for a walrus
```suggestion
        if (funcs := call_cuda_funcs.get(self.key.__name__)):
            for sig in funcs.keys():
                if all(
                    arg.group_scalar_type == ty for arg, ty in zip(args, sig)
                ):
                    return nb_signature(sig[0], *args)
```